### PR TITLE
8280881: (fs) UnixNativeDispatcher.close0 may throw UnixException

### PR DESCRIPTION
**Pull request by bplb**
Re-specify `sun.nio.fs.UnixNativeDispatcher.close0()` to throw `sun.nio.fs.UnixException` and modify callers either to ignore it or to rethrow it as an `IOException`, as appropriate.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280881](https://bugs.openjdk.java.net/browse/JDK-8280881): (fs) UnixNativeDispatcher.close0 may throw UnixException

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7467/head:pull/7467` \
`$ git checkout pull/7467`

Update a local copy of the PR: \
`$ git checkout pull/7467` \
`$ git pull https://git.openjdk.java.net/jdk pull/7467/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7467`

View PR using the GUI difftool: \
`$ git pr show -t 7467`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7467.diff">https://git.openjdk.java.net/jdk/pull/7467.diff</a>

</details>

### CODE REVIEW

**Review by LanceAndersen — COMMENTED** [review]
The changes look reasonable to throw an IOException vs the specific UnixException

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java [hunk 1]
```diff
@@ -431,10 +431,20 @@ public void createDirectory(Path obj, FileAttribute<?>... attrs)
...
         } catch (UnixException x) {
-            if (dfd1 != -1)
-                UnixNativeDispatcher.close(dfd1);
-            if (dfd2 != -1)
-                UnixNativeDispatcher.close(dfd2);
+            if (dfd1 != -1) {
+                try {
+                    UnixNativeDispatcher.close(dfd1);
+                } catch (UnixException y) {
+                    x.addSuppressed(y);
```
Comment (jaikiran) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:438]: Hello Brian,
I think, even after adding this as suppressed exception, this (suprressed) exception will be "lost" because at the end of this outer catch block there's this code:
```
x.rethrowAsIOException(dir);
```
and the implementation of `rethrowAsIOException` internally just creates a new exception instance (depending on the error code) but doesn't capture any suppressed exceptions from the original `x`.

File: src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java [hunk 2]
```diff
@@ -118,9 +122,17 @@ private UnixPath getName(Path obj) {
                 ptr = fdopendir(newdfd1);
             } catch (UnixException x) {
                 if (newdfd1 != -1)
-                    UnixNativeDispatcher.close(newdfd1);
+                    try {
+                        UnixNativeDispatcher.close(newdfd1);
+                    } catch (UnixException e) {
+                        x.addSuppressed(e);
```
Comment (jaikiran) [inline src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java:128]: Same here with respect to suppressed exception and what happens at the end of this catch block with `x.rethrowAsIOException(file);`

File: src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java [hunk 3]
```diff
@@ -133,7 +133,11 @@
                 null, "Unable to get list of extended attributes: " +
                 x.getMessage());
         } finally {
-            close(fd);
+            try {
+                close(fd);
+            } catch (UnixException y) {
+                y.rethrowAsIOException(file);
```
Comment (jaikiran) [inline src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java:139]: Hello Brian,
A few lines above this code in the finally block, there's a catch block which catches the `UnixException` and rethrows it as a `FileSystemException`. With the change being done in this finally block, this finally block now potentially can throw an `IOException` from a failed call to `close()` and thus the original `FileSystemException` can get lost. Do you think it would be better to keep track any potential `FileSystemException` from the previous catch block and add this `IOException` from this `close()` call as a suppressed exception to that `FileSystemException` and rethrow that original exception? Loosely, something like:

```
IOException failure = null;
try {
	....

} catch (UnixException x) {
	failure = new FileSystemException(file.getPathForExceptionMessage(),
	    null, "Unable to get list of extended attributes: " +
	    x.getMessage());
} finally {
    try {
	    close(fd);
	} catch (UnixException y) {
		if (failure == null) {
			// throw the exception that happened during close
			y.rethrowAsIOException(file);
		} else {
			// add the failure to close as a suppressed exception to
			// the previously caught failure
			failure.addSuppressed(y.asIOException(file));
		}
	}
	// throw any previously caught failure
	if (failure != null) {
		throw failure;
	}
}
```

This applies to a few others places that have changed in this PR.

**Comment by jaikiran** [comment]
Additionally the GitHub actions job on Linux system seems to have caught a few other places where the call to `close` now needs to catch the `UnixException`:

```
2022-02-14T21:24:09.6575143Z Compiling 3090 files for java.base
2022-02-14T21:24:37.0964078Z /home/runner/work/jdk/jdk/jdk/src/java.base/linux/classes/sun/nio/fs/LinuxDosFileAttributeView.java:185: error: unreported exception UnixException; must be caught or declared to be thrown
2022-02-14T21:24:37.0964759Z             close(fd);
2022-02-14T21:24:37.0964944Z                  ^
2022-02-14T21:24:37.0965316Z /home/runner/work/jdk/jdk/jdk/src/java.base/linux/classes/sun/nio/fs/LinuxDosFileAttributeView.java:280: error: unreported exception UnixException; must be caught or declared to be thrown
2022-02-14T21:24:37.0965679Z             close(fd);
2022-02-14T21:24:37.0965857Z                  ^
2022-02-14T21:24:37.0966014Z 2 errors
```

**Comment by bplb** [comment]
> Additionally the GitHub actions job on Linux system seems to have caught a few other places where the call to `close` now needs to catch the `UnixException`:

Indeed I had not addressed Linux yet.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java (hunk 1 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:438]: Good point.

File: src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java (hunk 2 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java:128]: Ditto.

File: src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java (hunk 3 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java:139]: This might be a good idea. I think I need to revisit all the changes in this patch.

Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java]: In commit 0e7b72f I elected to do the following:

1. If `close()` throws a `UnixException` in a finally block, then it is ignored unless it can be ascertained that it would would _not_ mask an exception thrown in the try block, in which case it is rethrown as an `IOException`.
2. The `UnixException` is also ignored if the calling method does not throw `IOException`.
3. If the `UnixException` can be added as a suppressed exception to a preexisting `IOException` then that is done.
4. Otherwise if the `UnixException` can be rethrown as an `IOException` without masking another exception than that is done.

File: src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java [hunk 4]
```diff
@@ -296,9 +301,12 @@ void implCloseAll() {
...
             // free resources
             unsafe.freeMemory(address);
-            UnixNativeDispatcher.close(socketpair[0]);
-            UnixNativeDispatcher.close(socketpair[1]);
-            UnixNativeDispatcher.close(ifd);
+            try {
+                UnixNativeDispatcher.close(socketpair[0]);
+                UnixNativeDispatcher.close(socketpair[1]);
+                UnixNativeDispatcher.close(ifd);
+            } catch (UnixException ignore) {
```
Comment (jaikiran) [inline src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java:308]: Hello Brian,
Do you think it would be better for each of these `close` calls to have a try/catch/ignore block of its own? That way, for example, if the close for `socketpair[0]` fails, then we at least attempt closing the other open ones (`socketpair[1]` and `ifd`)?

File: src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java (hunk 3 above)
Comment (jaikiran) [reply src/java.base/unix/classes/sun/nio/fs/UnixUserDefinedFileAttributeView.java]: Thank you for these changes, Brian. Overall these look fine to me. There's one part in this change which I'm unsure about given my lack of knowledge - the place where we throw an `IOException` depending on whether or not the operation is `done` or `complete`.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java [hunk 5]
```diff
@@ -209,8 +213,14 @@ private static void copyDirectory(UnixPath source,
...
             done = true;
         } finally {
-            if (dfd >= 0)
-                close(dfd);
+            if (dfd >= 0) {
+                try {
+                    close(dfd);
+                } catch (UnixException e) {
+                    if (done)
+                        e.rethrowAsIOException(target);
```
Comment (jaikiran) [inline src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:221]: Should we just ignore the exception irrespective of whether or not it is `done`? That I think would probably be a bit more consistent with some other places where we ignore exception from `close`. I think ignoring this instead of (conditionally) throwing this will also help propagate any previous raised exception, where a few lines above this `finally` block there's an "inner" try/catch which can potentially throw an `IOException` (line 211).

**Comment by bplb** [comment]
So in finally blocks where throwing an `IOException` because of a `close()` failure would mask any exception thrown in the `try` block it would be
```
close(fd, null);
```
which would mean ignoring the `UnixException` of `close()`?

**Comment by AlanBateman** [comment]
> So in finally blocks where throwing an `IOException` because of a `close()` failure would mask any exception thrown in the `try` block it would be
> 
> ```
> close(fd, null);
> ```
> 
> which would mean ignoring the `UnixException` of `close()`?

close(fd, e -> null) should work for the cases where it's okay to ignore. The important cases to throw is when writing but most of the cases will be reading.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java [hunk 6]
```diff
@@ -186,7 +186,7 @@ private static void copyDirectory(UnixPath source,
                 }
                 if (sfd >= 0) {
                     source.getFileSystem().copyNonPosixAttributes(sfd, dfd);
-                    close(sfd);
+                    close(sfd, x -> x.asIOException(source));
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:189]: This is the source fd so it's reading. It's not wrong to have it throw IOException, just a bit consistent to the other places where the exception is ignored.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java [hunk 7]
```diff
@@ -431,13 +431,18 @@ public void createDirectory(Path obj, FileAttribute<?>... attrs)
...
+            IOException ioe = x.errno() == UnixConstants.ENOTDIR ?
+                new NotDirectoryException(dir.getPathForExceptionMessage()) :
+                x.asIOException(dir);
+            if (dfd1 != -1) {
+                ioe = UnixNativeDispatcher.close(dfd1, ioe,
+                    y -> y.asIOException(dir));
+            }
+            if (dfd2 != -1) {
+                ioe = UnixNativeDispatcher.close(dfd2, ioe,
+                    z -> z.asIOException(dir));
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:443]: It's not clear to me that the closing of dfd1 and dfd2 are interesting here, maybe they should be e -> null too.

File: src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java [hunk 8]
```diff
@@ -117,13 +117,18 @@ private UnixPath getName(Path obj) {
...
+                    new NotDirectoryException(file.toString()) :
+                    x.asIOException(file);
+                if (newdfd1 != -1) {
+                    ioe = UnixNativeDispatcher.close(newdfd1, ioe,
+                        y -> y.asIOException(file));
+                }
+                if (newdfd2 != -1) {
+                    ioe = UnixNativeDispatcher.close(newdfd1, ioe,
+                        z -> z.asIOException(file));
+                }
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixSecureDirectoryStream.java:130]: Same as UnixFileSystemProvider, I don't think the errors from closing newfd1 and newfd2 are useful here.

File: src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java [hunk 9]
```diff
@@ -70,8 +70,8 @@
             socketpair(sp);
             configureBlocking(sp[0], false);
         } catch (UnixException x) {
-            UnixNativeDispatcher.close(ifd);
-            throw new IOException(x.errorString());
+            throw UnixNativeDispatcher.close(ifd,
+                new IOException(x.errorString()), y -> y.asIOException(null));
```
Comment (AlanBateman) [inline src/java.base/linux/classes/sun/nio/fs/LinuxWatchService.java:74]: The most likely cause of socketpair failing here would be resources or out of file descriptors. When that happens then it's probably not interesting to also try to report the failure of inotify close. So I think this close can be close(ifd, e-> null) too.

**Review by AlanBateman — COMMENTED** [review]
This looks much cleaner but it may be a bit overboard in a few places where there is an exception and the cleanup has to close and also fails.

File: src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java [hunk 10]
```diff
@@ -431,13 +431,14 @@ public void createDirectory(Path obj, FileAttribute<?>... attrs)
...
             if (dfd1 != -1)
-                UnixNativeDispatcher.close(dfd1);
+                ioe = UnixNativeDispatcher.close(dfd1, ioe, e -> null);
             if (dfd2 != -1)
-                UnixNativeDispatcher.close(dfd2);
-            if (x.errno() == UnixConstants.ENOTDIR)
-                throw new NotDirectoryException(dir.getPathForExceptionMessage());
-            x.rethrowAsIOException(dir);
+                ioe = UnixNativeDispatcher.close(dfd2, ioe, e -> null);
+            throw ioe;
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixFileSystemProvider.java:441]: I think it would be simpler to just revert the change and just call close(dfd1, e -> null) and close(dfd2, e -> null).

**Review by AlanBateman — COMMENTED** [review]
I think it's improving but I think can be simplified.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java [hunk 11]
```diff
@@ -186,7 +186,7 @@ private static void copyDirectory(UnixPath source,
                 }
                 if (sfd >= 0) {
                     source.getFileSystem().copyNonPosixAttributes(sfd, dfd);
-                    close(sfd);
+                    close(sfd, null);
```
Comment (AlanBateman) [inline src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java:189]: This should be `close(sfd, e -> null);`

**Review by AlanBateman — COMMENTED** [review]
This version is looking much better, just one issue in UnixCopyFile, the rest looks good.

File: src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java (hunk 11 above)
Comment (bplb) [reply src/java.base/unix/classes/sun/nio/fs/UnixCopyFile.java]: Fixed in commit 0bc1ffd.

**Review by AlanBateman — APPROVED** [review]
